### PR TITLE
Simulate larger batches

### PR DESCRIPTION
**Pull request by workofart**
## Description
To allow a limited memory machine to train with larger batch sizes, we are going to introduce a new functionality: `update_weights_every_n_steps`. This is to simulate larger batches by updating the weights once every N batches/steps.

**Changes:** Remove the `.step()` and `.zero_grad()` calls in the trainer class' `train_step()` function. We will be doing those two steps in the AbstractTrainer's `_train_one_epoch()`, which allows us to access the number of steps we have processed in a given epoch, and periodically perform `.step()` and `.zero_grad()` to achieve this.

Assume we have this training config in the example below: 
```
batch_size=16,
total_epochs=30,
steps_per_epoch=1600,
update_weights_every_n_steps=8, # simulate larger batch sizes
```
In each weight update, we will be processing 16 x 8 = 128 samples (simulating batch_size of 128, but we only keep 16 samples in the memory). Each epoch will run 1600 / 8 = 200 weight updates. For the entire training run, we will be doing 30 x 200 = 6000 weight updates.

## Tests
Updated one unit test to remove the `.step()` validation when the `train_step()` is called, because its reponsibility is no longer to perform the `.step()` function, explained above.

### CODE REVIEW

File: autograd/tools/trainer.py [hunk 1]
```diff
@@ -93,14 +92,21 @@ def fit(self, train_data_loader, val_data_loader=None):
 
             self._on_epoch_end(epoch, train_loss, val_loss)
 
-        self._save_metrics()
+            if epoch % (self.config.total_epochs // min(20, self.config.total_epochs)) == 0:
```
Comment (workofart) [inline autograd/tools/trainer.py:95]: Previously this was only saved at the very end, which doesn't make sense because we often early terminate the training run.

File: autograd/tools/trainer.py [hunk 2]
```diff
@@ -201,12 +206,18 @@ def _load_model_and_optimizer(
         """
         if resume_epoch is not None:
             # Look for checkpoint files
-            ckpt_json = os.path.join(
-                self.CHECKPOINT_DIR, f"{model_class.__name__}_{resume_epoch}.json"
-            )
-            ckpt_npz = os.path.join(
-                self.CHECKPOINT_DIR, f"{model_class.__name__}_{resume_epoch}.npz"
-            )
+            if checkpoint_path is not None:
```
Comment (workofart) [inline autograd/tools/trainer.py:209]: Making sure we properly handle the checkpoint path with the training_run_name prepended. The checkpoint saving logic was already already on line 107. This is a bug fix.

File: autograd/tools/trainer.py [hunk 3]
```diff
@@ -397,7 +406,7 @@ def _perform_inference(self, train_data_loader):
             prediction_func=self.forward_fn,
             bpe=train_data_loader.bpe,
             start_tokens=self.config.eval_start_string,
-            max_length=int(train_data_loader.seq_len * 0.4),
+            max_length=max(128, int(train_data_loader.seq_len * 0.4)),
```
Comment (workofart) [inline autograd/tools/trainer.py:409]: When our context length is long, this spams the terminal logs. Keeping it relatively short for a balance of readability and informative.

File: test/autograd/tools/test_trainer.py [hunk 4]
```diff
@@ -167,7 +167,6 @@ def test_train_step_returns_loss(self):
         batch = next(iter(self.train_data))
         loss_val = self.trainer.train_step(batch)
         self.assertAlmostEqual(loss_val, 1.23, places=5)
-        self.assertEqual(self.trainer.optimizer.step_call_count, 1)
```
Comment (workofart) [inline test/autograd/tools/test_trainer.py:170]: Described in the PR description.

File: autograd/tools/trainer.py [hunk 5]
```diff
@@ -159,7 +171,6 @@ def _on_epoch_end(self, epoch: int, train_loss: float, val_loss: Optional[float]
             if key != "epoch" and self.metrics[key][-1] is not None:
                 log_msg += f"\n{key} = {self.metrics[key][-1]:.4f}"
         log_msg += f"\nLR = {self.optimizer.lr:.4f}"
-        log_msg += f"\nGrad L2 Norm = {grad_l2_norm(self.model.parameters):.2f}"
```
Comment (workofart) [inline autograd/tools/trainer.py:162]: Moved to line 112, since we have to compute the gradients before we call `zero_grad()`
